### PR TITLE
SWC-36: changes for markdown in Description support

### PR DESCRIPTION
**Pull request by jay-hodgson**

### CODE REVIEW

File: src/main/java/org/sagebionetworks/web/client/DisplayUtils.java [hunk 1]
```diff
@@ -1049,4 +1051,44 @@ public static Anchor createIconLink(AbstractImagePrototype icon, ClickHandler cl
 		return anchor;
 	}
 
+	/**
+	 * During the rendering process, the markdown is converted to html, then postprocessed.  This is the postprocessing step.
+	 * It adds the given css classname to entities supported by the markdown. also detects Synapse IDs, and creates links out of them
+	 * @param panel
+	 */
+	public static String postProcessEntityDescriptionHtml(String html, String cssClassName){
```
Comment (dburdick) [inline src/main/java/org/sagebionetworks/web/client/DisplayUtils.java]: Maybe show attachment links with their own special link syntax and resolve that into the full resolvable URL in post processing. something like:

Attachment://entity/123/token/456

File: src/main/java/org/sagebionetworks/web/client/widget/entity/EntityPageTopViewImpl.java [hunk 2]
```diff
@@ -623,59 +626,6 @@ private Widget createPropertyWidget(EntityBundle bundle) {
 		return lc;
 	}
 
-	/**
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/widget/entity/EntityPageTopViewImpl.java:626]: add back in

File: src/main/java/org/sagebionetworks/web/client/widget/entity/EntityPropertyForm.java [hunk 3]
```diff
@@ -53,15 +77,25 @@
...
+	SynapseClientAsync synapseClient;
+	SafeHtml showFormattingTipsSafeHTML, hideFormattingTipsSafeHTML;
+	
 	@Inject
-	public EntityPropertyForm(FormFieldFactory formFactory, IconsImageBundle sageImageBundle) {
+	public EntityPropertyForm(FormFieldFactory formFactory, IconsImageBundle iconsImageBundle, SynapseClientAsync synapseClient) {
 		this.formFactory = formFactory;
-		this.iconsImageBundle = sageImageBundle;
+		this.iconsImageBundle = iconsImageBundle;
+		this.synapseClient = synapseClient;
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/widget/entity/EntityPropertyForm.java:96]: pull synapse client call (for preview) out of this widget

File: src/main/java/org/sagebionetworks/web/client/widget/entity/EntityPropertyForm.java [hunk 4]
```diff
@@ -212,17 +335,64 @@ public void rebuild() {
 	}
 	
 	/**
+	 * Show the edit entity dialog.
+	 * @param entity
+	 * @param annos
+	 * @param callback
+	 */
+	public static void showPreviewWindow(SynapseClientAsync synapseClient, String descriptionMarkdown){
```
Comment (dburdick) [inline src/main/java/org/sagebionetworks/web/client/widget/entity/EntityPropertyForm.java:343]: Extract Synapse client and preview generation to a presenter (perhaps via an interface)

File: src/main/java/org/sagebionetworks/web/client/widget/entity/dialog/VisualAttachmentsList.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+import com.google.inject.Inject;
+
+public class VisualAttachmentsList implements VisualAttachmentsListView.Presenter,
+		SynapseWidgetPresenter {
+
+	private VisualAttachmentsListView view;
+	private List<AttachmentData> attachments;
+	
+	@Inject
+	public VisualAttachmentsList(VisualAttachmentsListView view, SynapseClientAsync synapseClient){
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/widget/entity/dialog/VisualAttachmentsList.java:20]: remove synapseClient!

File: src/main/java/org/sagebionetworks/web/client/widget/entity/EntityPropertyForm.java [hunk 6]
```diff
@@ -191,15 +228,101 @@ public void rebuild() {
...
+			@Override
+			public void componentSelected(ButtonEvent ce) {
+				//pop up a list of attachments, and have the user pick one.
+				SelectAttachmentDialog.showSelectAttachmentDialog(baseURl,  entityId, attachments, synapseClient, "Select Attachment", "Insert", new SelectAttachmentDialog.Callback() {
+					
+					@Override
+					public void onSelectAttachment(AttachmentData data) {
+						//insert the markdown into the description for the image attachment
+						SafeHtml safeName = SafeHtmlUtils.fromString(data.getName());
+						StringBuilder sb = new StringBuilder();
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/widget/entity/EntityPropertyForm.java:289]: pull out into utility class, and then test

File: src/main/java/org/sagebionetworks/web/client/widget/entity/EntityPropertyForm.java [hunk 7]
```diff
@@ -212,17 +335,64 @@ public void rebuild() {
...
+	    		}
+	    		FlowPanel f = new FlowPanel();
+	    		f.setStyleName("entity-description-preview-wrapper");
+	    		f.add(panel);
+	    		window.add(new ScrollPanel(f));
+				window.show();
+			}
+			@Override
+			public void onFailure(Throwable caught) {
+				//preview failed
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/client/widget/entity/EntityPropertyForm.java:376]: pop up error message notifying that preview failed

File: src/main/java/org/sagebionetworks/web/server/servlet/SynapseClientImpl.java [hunk 8]
```diff
@@ -874,4 +887,191 @@ public EntityWrapper updateExternalLocationable(String entityId, String external
...
+		    	}
+			}
+		}
+	}
+
+	public static void addSynapseLinks(Document doc) {
+		// in this case, I still need a regular expression to find the synapse ids.
+		// find all elements whose text contains a synapse id pattern
+		// replace the TextNode element children with Elements, whose html contain a link to relevant synapse entity.
+		String regEx = "\\W*(syn\\d+)\\W*";
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/server/servlet/SynapseClientImpl.java:957]: describe

File: src/main/java/org/sagebionetworks/web/server/servlet/SynapseClientImpl.java [hunk 9]
```diff
@@ -874,4 +887,191 @@ public EntityWrapper updateExternalLocationable(String entityId, String external
...
+				Node childNode = (Node) iterator2.next();
+				if (childNode instanceof TextNode) {
+					String oldText = ((TextNode) childNode).text();
+					// find it in the text
+					Matcher matcher = pattern.matcher(oldText);
+					StringBuilder sb = new StringBuilder();
+					int previousFoundIndex = 0;
+					while (matcher.find() && matcher.groupCount() == 1) {
+						sb.append(oldText.substring(previousFoundIndex, matcher.start(1)));
+						sb.append(getSynAnchorHtml(matcher.group(1))); //the actual synapse Id group (not the non-word characters that might surround it)
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/server/servlet/SynapseClientImpl.java:973]: what happens when I have a synapse ID in the link text?

File: src/main/java/org/sagebionetworks/web/server/servlet/SynapseClientImpl.java [hunk 10]
```diff
@@ -874,4 +887,191 @@ public EntityWrapper updateExternalLocationable(String entityId, String external
...
+					newElement.html(sb.toString());
+					childNode.replaceWith(newElement);		
+				}
+			}
+		}
+	}
+	
+	public static void addYouTubeVideos(Document doc) {
+		// using a regular expression to find our special YouTube video embedding notation, replace with an embedded version of the video!
+		String regEx = "\\W*(\\{youtube=(\\w*)\\})\\W*";
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/server/servlet/SynapseClientImpl.java:989]: attempt to pull out, this pattern (of patterns) will be reused

File: src/test/java/org/sagebionetworks/web/unitserver/SynapseClientImplTest.java [hunk 11]
```diff
@@ -428,4 +433,80 @@ public void testUpdateLocationable() throws Exception {
...
+		String expectedResult = "<html>\n <head></head>\n <body>\n  <ul class=\" myclass\">\n   <li>Abacus\n    <ul class=\" myclass\">\n     <li>answer</li>\n    </ul></li>\n   <li>Bubbles\n    <ol class=\" myclass\">\n     <li>bunk</li>\n     <li>bupkis\n      <ul class=\" myclass\">\n       <li>BELITTLER</li>\n      </ul></li>\n     <li>burper</li>\n    </ol></li>\n   <li>Cunning</li>\n  </ul>\n  <blockquote class=\" myclass\"> \n   <p>Email-style angle brackets are used for blockquotes.</p>\n  </blockquote> \n  <p><code>&lt;code&gt;</code> spans are delimited by backticks.</p>\n  <p>An <a href=\"http://url.com/\" title=\"Title\" class=\" myclass\">example</a></p>\n </body>\n</html>";
+		Document htmlDoc = Jsoup.parse(testString);
+		SynapseClientImpl.applyCssClass(htmlDoc, "myclass");
+		String actualResult = htmlDoc.html();
+		assertEquals(expectedResult, actualResult);
+	}
+
+	@Test
+	public void testImageAttachmentLinks(){
+		StringBuilder sb = new StringBuilder();
```
Comment (jay-hodgson) [inline src/test/java/org/sagebionetworks/web/unitserver/SynapseClientImplTest.java:480]: use util method to form attachment link

File: src/main/webapp/Portal.css [hunk 12]
```diff
@@ -454,6 +454,29 @@ a.shortBtn {
...
+	list-style-type: decimal;
+	padding-left: 17px;
+}
+ul.from-markdown { 
+	list-style-type: square;
+	padding-left: 17px;
+}
+strong.from-markdown { font-weight: bold; }
+em.from-markdown { font-style: italic; }
+a.auto-detected-synapse-link, a.auto-detected-synapse-link:hover, a.auto-detected-synapse-link:visited { color:#BC3226;}
```
Comment (jay-hodgson) [inline src/main/webapp/Portal.css:467]: remove color definitions here, apply link class to anchors in code

File: src/test/java/org/sagebionetworks/web/unitclient/widget/entity/EntityPageTopTest.java [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+		final String testMarkdown = "HTML Returns\n----------";
+		AsyncMockStubber.callSuccessWith(testHtml).when(mockSynapseClient).markdown2Html(any(String.class), any(String.class), any(AsyncCallback.class));
+		AsyncCallback<String> callback = new AsyncCallback<String>() {
+			@Override
+			public void onSuccess(String result) {
+				assertEquals("unexpected markdown to html conversion", testHtml, result);
+			}
+			@Override
+			public void onFailure(Throwable caught) {
+				throw new IllegalArgumentException(caught);
```
Comment (jay-hodgson) [inline src/test/java/org/sagebionetworks/web/unitclient/widget/entity/EntityPageTopTest.java:78]: use junit fail method (pass at least the exception message)

**Pull request merged**